cdap-watchdog/src/main/java/co/cask/cdap/metrics/store/upgrade/DataMigration.java

```diff
@@ -0,0 +1,329 @@
+    if (contextParts.size() > 0) {
+      systemMap = mapOldSystemContextToNew.get(contextParts.get(0));
+    }
+    String tagKey = metricNameToTagNameMapping.get(metricName);
+    if (systemMap != null) {
+      tagMap.putAll(systemMap);
+    } else  if (contextParts.size() > 1) {
+      // application metrics
+      List<String> targetTagList = typeToTagNameMapping.get(contextParts.get(1));
+      if (targetTagList != null) {
+        populateApplicationTags(tagMap, contextParts, targetTagList, context);
+      } else {
+        // service metrics , we are skipping them as they were not exposed for querying before
+        LOG.trace("Skipping context : {}", context);
+        return;
+      }
+    } else {
+      LOG.warn("Should not reach here {}", context);
+    }
+    LOG.trace("Adding metrics - tagMap : {} - context : {} - metricName : {} and tagKey : {}", tagMap, context,
+              metricName, tagKey);
+    addMetrics(iterator, metricName, tagMap, tagKey);
+  }
+
+  private void populateApplicationTags(Map<String, String> tagMap, List<String> contextParts,
+                                       List<String> targetTagList, String context) {
+    tagMap.put(Constants.Metrics.Tag.NAMESPACE, Constants.DEFAULT_NAMESPACE);
+    for (int i = 0; i < contextParts.size(); i++) {
+      if (i == targetTagList.size()) {
+        LOG.info(" Context longer than targetTagList" + context);
+        break;
+      }
+      if (targetTagList.get(i).equals(TYPE)) {
+        continue;
+      }
+      tagMap.put(targetTagList.get(i), contextParts.get(i));
+    }
+  }
+
+  // constructs MetricValue based on parameters passed and adds the MetricValue to MetricStore.
+  private void addMetricValueToMetricStore(Map<String, String> tags, String metricName,
+                                           int timeStamp, long value, MetricType counter) throws Exception {
+    LOG.info("Storing Metric - Tags {} metricName {} value {}", tags, metricName, value);
+    aggMetricStore.add(new MetricValue(tags, metricName, timeStamp, value, counter));
+  }
+
+  private MetricsTable getOrCreateMetricsTable(String tableName, DatasetProperties empty) {
+    LOG.info("Get Metrics Table" + tableName);
+    MetricsTable table = null;
+    // metrics tables are in the system namespace
+    Id.DatasetInstance metricsDatasetInstanceId = Id.DatasetInstance.from(Constants.SYSTEM_NAMESPACE, tableName);
+    try {
+      table = DatasetsUtil.getOrCreateDataset(dsFramework, metricsDatasetInstanceId,
+                                              MetricsTable.class.getName(), empty, null, null);
+    } catch (DatasetManagementException e) {
+      LOG.error("Cannot access or create table {}.", tableName);
+    } catch (IOException e) {
+      LOG.error("Exception while creating table {}.", tableName, e);
+      throw Throwables.propagate(e);
+    }
+    return table;
+  }
+
+  private String getMetricNameBasedOnVersion(String metricName, Version version) {
```

Comment:
if you call it on every record, it looks really expensive. Can we use some NameResolver interface with 2.6 and 2.7 implementations to be passed around? or smth like that